simple-ipc--unix-sockets.c

```diff
@@ -170,4 +170,898 @@ int ipc_client_send_command(const char *path,
+ * up correctly.
+ *
+ * When future service instances try to create the listener socket,
+ * `bind(2)` will fail with EADDRINUSE -- because the inode already
+ * exists.  However, the new instance cannot tell if it is a stale
+ * inode *or* another service instance is already running.
+ *
+ * One possible solution is to blindly unlink the inode before
+ * attempting to create the new socket and inode.  Then `bind(2)`
+ * should always succeed.  However, if there is an existing service
+ * instance, it would be orphaned -- listening on a socket/inode that
+ * is no longer visible to future clients.
+ *
+ * An alternative solution is to gently create the socket.  If the
+ * inode already exists, try to determine whether it is stale or
+ * actually in use by an existing service process by attempting to
+ * connect to the socket/inode.  (This assumes that the existing
+ * process isn't wedged.)  Then if stale, recreate the socket with
+ * force.
+ */ 
+static int create_listener_socket(const char *path)
+{
+	int fd_listen;
+	int fd_client;
+	struct unix_stream_listen_opts opts = {
+		.listen_backlog_size = LISTEN_BACKLOG,
+		.force_unlink_before_bind = 0,
+		.disallow_chdir = 1
+	};
+
+	trace2_data_string("ipc-server", NULL, "try-listen-gently", path);
+
+	fd_listen = unix_stream_listen_gently(path, &opts);
+	if (fd_listen >= 0)
+		return fd_listen;
+
+	if (errno != EADDRINUSE)
+		return error_errno(_("could not create socket '%s'"),
+				   path);
+
+	trace2_data_string("ipc-server", NULL, "try-detect-server", path);
+
+	fd_client = unix_stream_connect(path);
+	if (fd_client >= 0) {
+		/*
+		 * An existing service process is alive and accepted our
+		 * connection.
+		 */
+		close(fd_client);
+
+		errno = EADDRINUSE;
+		return error_errno(_("socket already in use '%s'"),
+				   path);
+	}
+
+	trace2_data_string("ipc-server", NULL, "try-listen-force", path);
+
+	opts.force_unlink_before_bind = 1;
+	fd_listen = unix_stream_listen_gently(path, &opts);
+	if (fd_listen >= 0)
+		return fd_listen;
+
+	return error_errno(_("could not force create socket '%s'"), path);
+}
+
+static int setup_listener_socket(const char *path, ino_t *inode)
+{
+	int fd_listen;
+	struct stat st;
+
+	trace2_region_enter("ipc-server", "create-listener_socket", NULL);
+	fd_listen = create_listener_socket(path);
+	trace2_region_leave("ipc-server", "create-listener_socket", NULL);
+
+	if (fd_listen < 0)
+		return fd_listen;
+
+	/*
+	 * We just bound a socket (descriptor) to a newly created unix
+	 * domain socket in the filesystem.  Capture the inode number
+	 * so we can later detect if/when someone else force-creates a
+	 * new socket and effectively steals the path from us.  (Which
+	 * would leave us listening to a socket that no client could
+	 * reach.)
+	 */
+	if (lstat(path, &st) < 0) {
+		int saved_errno = errno;
+
+		close(fd_listen);
+		unlink(path);
+
+		errno = saved_errno;
+		return error_errno(_("could not lstat listener socket '%s'"),
+				   path);
```

Comment:
Why not call `error_errno()` _before_ closing and unlinking, and then finally `return -1`?

Reply:
i wanted the caller to have a valid errno in some cases, but I don't remember the details.

Reply:
Oh, that makes sense!